libbpf-tools: Add new feature ALSan

Bojun-Seo opened the pull request:
Add ALSan(Attachable Leak Sanitizer) feature on libbpf-tools
ALSan feature originally comes from the llvm-project lsan
https://github.com/llvm/llvm-project
cvector.h comes from c-vector project, `commit d3f3156373b0587336ac7ee1568755d6cf93dd39`
https://github.com/eteran/c-vector
uthash.h comes from uthash project, `commit bf15263081be6229be31addd48566df93921cb46`
https://github.com/troydhanson/uthash
This tool detect and report unreachable memory periodically


Usage:
```
$ ./alsan -h
Usage: alsan [OPTION...]
Detect memory leak resulting from unreachable pointers.

Either -c or -p is a mandatory option
EXAMPLES:
    alsan -p 1234             # Detect leaks on process id 1234
    alsan -c a.out            # Detect leaks on a.out
    alsan -c 'a.out arg'      # Detect leaks on a.out with argument

  -c, --command=COMMAND      Execute and detect memory leak on the specified
                             command
  -i, --interval=INTERVAL    Set interval in second to detect leak
  -p, --pid=PID              Detect memory leak on the specified process
  -s, --suppressions=SUPPRESSIONS
                             Suppressions file name
  -T, --top=TOP              Report only specified amount of backtraces
  -v, --verbose              Verbose debug output
  -w, --stop-the-world       Stop the target process during tracing
  -?, --help                 Give this help list
      --usage                Give a short usage message
  -V, --version              Print program version

Mandatory or optional arguments to long options are also mandatory or optional
for any corresponding short options.

Report bugs to https://github.com/iovisor/bcc/tree/master/libbpf-tools.
```

Report example:
```
$ sudo ./alsan -p 28346

[2024-05-22 14:44:58] Print leaks:
44 bytes direct leak found in 1 allocations from stack id(57214)
        #1 0x00583bca1b2250 baz+0x1c (/home/bojun/alsan/libbpf-tools/a.out+0x1250)
        #2 0x00583bca1b22d7 main+0x73 (/home/bojun/alsan/libbpf-tools/a.out+0x12d7)
        #3 0x007470c7c2a1ca [unknown] (/usr/lib/x86_64-linux-gnu/libc.so.6+0x2a1ca)
        #4 0x007470c7c2a28b __libc_start_main+0x8b (/usr/lib/x86_64-linux-gnu/libc.so.6+0x2a28b)
        #5 0x00583bca1b2105 _start+0x25 (/home/bojun/alsan/libbpf-tools/a.out+0x1105)


[2024-05-22 14:45:08] Print leaks:
132 bytes direct leak found in 3 allocations from stack id(57214)
        #1 0x00583bca1b2250 baz+0x1c (/home/bojun/alsan/libbpf-tools/a.out+0x1250)
        #2 0x00583bca1b22d7 main+0x73 (/home/bojun/alsan/libbpf-tools/a.out+0x12d7)
        #3 0x007470c7c2a1ca [unknown] (/usr/lib/x86_64-linux-gnu/libc.so.6+0x2a1ca)
        #4 0x007470c7c2a28b __libc_start_main+0x8b (/usr/lib/x86_64-linux-gnu/libc.so.6+0x2a28b)
        #5 0x00583bca1b2105 _start+0x25 (/home/bojun/alsan/libbpf-tools/a.out+0x1105)

^C
$
```

Source code of test program:
```c
$ cat leak_test.c
#include <stdlib.h>
#include <unistd.h>

int *arr[10000];

int *foo(size_t size) {
        int *tmp = malloc(size);
        *tmp = 99;
        return tmp;
}

int *bar(size_t nmemb, size_t size) {
        int *tmp = calloc(nmemb, size);
        *tmp = 22;
        return tmp;
}

int *baz(size_t size) {
        int *tmp = valloc(size);
        *tmp = 11;
        return tmp;
}

int main(int argc, char* argv[]) {
        int *a;
        int i = 0;
        while (1) {
                a = foo(4);
                arr[i++] = a;
                a = bar(4, 4);
                free(a);
                a = baz(44);
                sleep(5);
        }
        return 0;
}
```

chenhengqi commented:
There is a similar PR in #3828 .

Bojun-Seo commented:
@chenhengqi, well, lsan is pretty different from memleak.
memleak is to keep tracking currently allocated heap memory,
where lsan is to report dangling pointers within "currently allocated heap memory" like llvm-project leak sanitizer

chenhengqi commented:
Fine, could you describe how does it work ?

Bojun-Seo commented:
Followings are the way it work.(Refer to this document https://github.com/google/sanitizers/wiki/AddressSanitizerLeakSanitizerDesignDocument)
lsan hooks memory allocating functions(malloc, calloc, free.. etc).
It considers memory space as direct graph.(Consider every value as a pointer)
Set hooked node as "unreachable", before it starts following steps.(and repeat periodically)
One, it traverses from the root set and mark reachable node as "reachable"
Two, it traverses again from the hooked set and marks reachable node as "indirect" if it is "unreachable" node.
Three, sort and report "unreachable" node as direct leak and "indirect" node as indirect leak.

About root set:
Root set denotes these things.
global variables, stacks of running threads, general-purpose registers of running threads, ELF thread-local storage and POSIX thread-specific data.
Original lsan is a library, so it can use API to get this information.
BPF lsan is an independent process, so it use ptrace api to get general-purpose registers, and read /proc/pid/maps to get others.

Indirect leak means not reachable allocation node from the root but can be reachable from some unreachable block.
Direct leak means not reachable allocation node from the root and not indirect leak.

Bojun-Seo commented:
The memory leak problem is an old problem and has not yet been conquered but very important. To solve memory leak problem, various methods have been tried and the best way is using a tool so far. There are two types of tools: static tools and dynamic tools.
In case of static tools, operation time and detection capability are trade off. Theoretically, almost infinite amount of time is required to detect all memory leaks using a static analysis tool. So static tool usually used to detect quite easy memory leaks because of time limitations.
In case of dynamic tools, there are two famous dynamic memory leak detecting tools: Valgrind and llvm-project LeakSanitizer.
These tools are good but have limitations. Valgrind performance degradation is about 10-50 times. It is difficult to use even in a desktop environment, and it is virtually impossible to use in an embedded system. llvm-project LeakSanitizer has little performance degradation, but not easy to use. And both tool require to restart the target program and can get report at termination of the target process.
I suggest bcc lsan(this PR) to solve these problems. It has little performance degradation, is easy to use, can be applied immediately to a running process, and can get report without target process termination.
For example, imagine if you developed a new service(daemon), which has no scenario to be quit. Somehow you found some weird memory increasement on the service. How can you debug it? To use llvm-project LeakSanitizer or Valgrind, you need to make exit routine on your program, regenerate the issue and rerun the program. Making good exit routine is quite hard but let's consider you did it well. But what if it takes 3 days to regenerate this issue? Maybe you need to stop the factory 3 days.
bcc lsan will reduce and remove these kind of problems. Because it can be applied immediately and generate report in the middle of the target process execution.

eiffel-fl reviewed (commented):
Hi.

First of all, I would like to thank you for your contribution, it seems really cool.
I have some comments regarding the eBPF code.
I will take a look later at the C userland code (even though I am not 100% I will be able to understand the logic).

Best regards.

Bojun-Seo commented on libbpf-tools/lsan.bpf.c:
```diff
@@ -0,0 +1,220 @@
+{
+	u64 pid = bpf_get_current_pid_tgid();
+	u64 size64 = size;
+	bpf_map_update_elem(&sizes, &pid, &size64, BPF_OK);
```
First of all, I'm so glad to say thanks to you in here again!

I will change from BPF_OK to BPF_ANY after check it works on my target(raspberry pi)
And you are right, I'm tracing one allocation from a given PID at a time.

Bojun-Seo replied:
First of all, I should have to use correct notation... But I didn't. Sorry to confuse you.
`bpf_get_current_pid_tgid` returns (user space) thread id.
(even though the name of this function looks like to returning process id)
https://github.com/iovisor/bcc/blob/master/docs/reference_guide.md#4-bpf_get_current_pid_tgid
So same thread cannot execute this code at the same time.

The answer of your question:
It saves `size` on allocation start for each thread and delete it on allocation return.
So even several allocation happens sequentially, no previous allocation history of `size` exists on `sizes` map.
`lsan` want to save the address of `malloc` function returns and the `size` value.
But 'size` is given on allocation start. And `malloc` can be executed on the other threads.
So this code saves `size` for each thread and use it on returning `malloc` function.
Thanks for good question!

Bojun-Seo commented on libbpf-tools/lsan.bpf.c:
```diff
@@ -0,0 +1,220 @@
+{
+	u64 addr = (u64)address;
+	struct lsan_info_t *info = bpf_map_lookup_elem(&allocs, &addr);
+	if (info == NULL)
+		return 0;
+
+	bpf_map_delete_elem(&allocs, &addr);
+	return 0;
```
Your suggestion and my code work equally, I guess.
But your code seems bit easy to be read.
So I'll change the code as you suggested.

Bojun-Seo commented on libbpf-tools/lsan.bpf.c:
```diff
@@ -0,0 +1,220 @@
+{
+	u64 memptr64 = (u64)(size_t)memptr;
+	u64 pid = bpf_get_current_pid_tgid();
+	bpf_map_update_elem(&memptrs, &pid, &memptr64, BPF_OK);
```
Same as the above comment, I'll change BPF_OK into BPF_ANY
Thanks again!

eiffel-fl reviewed (commented):
Hi.

This time, I took a look at the userspace code.
I am not 100% sure I understand the whole code particularly all the hash maps you need but it worked!
I also share your opinion about being able to quickly spot memory leak at runtime while not slowing the application (contrary to `valgrind` even though I really like this tool).

I tried to run it and I had to modify the following:

```patch
diff --git a/libbpf-tools/lsan.c b/libbpf-tools/lsan.c
index dcb3862e..1cd9538b 100644
--- a/libbpf-tools/lsan.c
+++ b/libbpf-tools/lsan.c
@@ -358,11 +358,11 @@ static int get_libc_path(char *path)
                        continue;
 
                file_name = strrchr(buf, '/') + 1;
-#ifdef __aarch64__
+// #ifdef __aarch64__
                if (sscanf(file_name, "libc-%f.so", &version) == 1) {
-#else
-               if (sscanf(file_name, "libc.so.%f", &version) == 1) {
-#endif
+// #else
+//             if (sscanf(file_name, "libc.so.%f", &version) == 1) {
+// #endif
                        memcpy(path, buf, strlen(buf));
                        fclose(f);
                        return 0;
```

Indeed, on my system, I have:

```bash
francis@pwmachine:~/Codes/kinvolk/bcc/libbpf-tools$ lsb_release -a                                                                                  
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 20.04.5 LTS
Release:        20.04
Codename:       focal
francis@pwmachine:~/Codes/kinvolk/bcc/libbpf-tools$ uname -a                                                                                        
Linux pwmachine 5.15.0-48-generic #54~20.04.1-Ubuntu SMP Thu Sep 1 16:17:26 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
francis@pwmachine:~/Codes/kinvolk/bcc/libbpf-tools$ ls -l /usr/lib/x86_64-linux-gnu/libc.so.6                                                       
lrwxrwxrwx 1 root root 12 avril  7 03:24 /usr/lib/x86_64-linux-gnu/libc.so.6 -> libc-2.31.so
```

After this modification, I was able to get the same output as you (plus the value 99 printed by `a.out`).


Best regards.

Bojun-Seo commented on libbpf-tools/cvector.h:
```diff
@@ -0,0 +1,198 @@
+
```
You are right, I should have to.
I add commit hash number on the description.
And I'm thinking about to use other project by adding submodule. What do you think?

Bojun-Seo replied:
I just copied only one file from each project.
So I agree with your opinion(getting file rather than using submodule).
And you said, I need to wait for maintainers to decide.

Bojun-Seo commented on libbpf-tools/lsan.c:
```diff
@@ -0,0 +1,1377 @@
+		exit(-1);
+	}
+
+	if (child_pid > 0) {
+		free(argv);
+		return child_pid;
+	}
```
You are right, I'll change it. Thanks!

Bojun-Seo commented on libbpf-tools/lsan.c:
```diff
@@ -0,0 +1,1377 @@
+	}
+}
+
+static uptr get_reference(uptr pp, size_t size)
```
Right, this function reads an address from another virtual memory space.
I think your suggested `dereference` is better name, I'll change it.

Bojun-Seo commented on libbpf-tools/lsan.c:
```diff
@@ -0,0 +1,1377 @@
+	struct iovec io;
+	long ret = 0;
+	int i = 0;
+	int j __attribute__((unused)) = 0;
```
I've been thinking about to use `#ifdef` as you said.
But I think it would be better to use `__attribute__((unused))` because other architecture may use this `j` value as well.

Bojun-Seo commented on libbpf-tools/lsan.c:
```diff
@@ -0,0 +1,1377 @@
+			scan_range_for_pointers(regs.regs[j],
+				regs.regs[j] + word_size, REACHABLE);
+		}
+#else // TODO: x86_64 implementation is needed
```
Great! thanks!
Can I use this code into my patch? I mean, isn't it your code?
Maybe you can contribute this code.

I should study x86_64 architecture register's usage, to be clear which registers are of interest.

Bojun-Seo commented on libbpf-tools/lsan.c:
```diff
@@ -0,0 +1,1377 @@
+	size_t i = 0;
+	size_t j = 0;
+
+	memset(report_buf, 0, BUF_MAX);
```
You are right, I'll delete calling `memset` in here.

Bojun-Seo commented on libbpf-tools/lsan.c:
```diff
@@ -0,0 +1,1377 @@
+const char *heap_str = "[heap]";
+const char *stack_str = "[stack]";
+// TODO: need to get this value from the system
+const size_t word_size = 8; // currently support 64-bit machine only
```
I'll change this code as you suggested. Thanks for good suggestion again.
I didn't sure about that `sizeof(void*)` value decided on host architecture or target architecture, in cross compile environment.
I tested it and found that it depends on target architecture.

chenhengqi commented:
Sorry for the long delay, I will try finishing this review this week.

Bojun-Seo commented on libbpf-tools/cvector.h:
```diff
@@ -0,0 +1,198 @@
+
```
Thanks for your review!
I ask to add copyright and license on the top of the cvector.h file
https://github.com/eteran/c-vector/pull/42
If this request is not accepted, I'll add the copyright manually

Bojun-Seo replied:
I update `cvector.h` file to the latest version, since https://github.com/eteran/c-vector/pull/42 merged.

Bojun-Seo commented on libbpf-tools/Makefile:
```diff
@@ -41,6 +41,7 @@ APPS = \
 	hardirqs \
 	klockstat \
 	ksnoop \
+	lsan \
```
I'll add it to `.gitignore`
And It seems lsan should go right after `llcstat`, I'll change the location either

Bojun-Seo commented on libbpf-tools/lsan.h:
```diff
@@ -0,0 +1,21 @@
+#define MAX_ENTRIES	65536
+
+enum chunk_tag {
+	DIRECTLY_LEAKED = 0,  // default
```
I'll change every `//` comments into `/*` and `*/` style

Bojun-Seo commented on libbpf-tools/lsan.bpf.c:
```diff
@@ -0,0 +1,219 @@
+
+
+char _license[] SEC("license") = "GPL";
+__u32 _version SEC("version") = 1;
```
I think I don't need `version` section.
I'll remove it.

Bojun-Seo commented on libbpf-tools/lsan.bpf.c:
```diff
@@ -0,0 +1,219 @@
+int gen_alloc_enter(struct pt_regs *ctx, size_t size)
+{
+	u64 pid = bpf_get_current_pid_tgid();
+	u64 size64 = size;
```
Is this rule only for `.bpf.c`? or for other `.c` file as well?

Bojun-Seo replied:
I changed `.c` files as well

Bojun-Seo commented on libbpf-tools/lsan.bpf.c:
```diff
@@ -0,0 +1,219 @@
+	__uint(key_size, sizeof(u32));
+} stack_traces SEC(".maps");
+
+static __always_inline
+int gen_alloc_enter(struct pt_regs *ctx, size_t size)
```
I'll change it as you said

chenhengqi commented:
Ping me if this PR is ready for review.

Bojun-Seo commented:
I update lsan to use following(merged) feature to show offsets of symbol and module.
https://github.com/iovisor/bcc/pull/4288

woodpenker commented:
I happen to use `uthash` before, just wondering if it can be replaced by the hashmap from `libbpf/src/hashmap.h`, so we don't need to introduce a new depend. Thanks.

Bojun-Seo commented:
I'll check whether `libbpf/src/hashmap.h` can be used instead of `uthash` in terms of implementation as well as license issue. Thanks for good review. By the way, what do you think about to using `C++` instead of `C`. All dependence(on `cvector` and `uthash`) can be removed easily, if I could use `C++`.

woodpenker commented:
> I'll check whether `libbpf/src/hashmap.h` can be used instead of `uthash` in terms of implementation as well as license issue. Thanks for good review. By the way, what do you think about to using `C++` instead of `C`. All dependence(on `cvector` and `uthash`) can be removed easily, if I could use `C++`.

Sorry, I have no idea whether the community is considered to use C++.

Bojun-Seo commented:
> > > I'll check whether `libbpf/src/hashmap.h` can be used instead of `uthash` in terms of implementation as well as license issue. Thanks for good review. By the way, what do you think about to using `C++` instead of `C`. All dependence(on `cvector` and `uthash`) can be removed easily, if I could use `C++`.
> > 
> > 
> > Sorry, I have no idea whether the community is considered to use C++.
> 
> For C++ I have also no idea, at one moment I wanted to use Rust because I had a similar issue than you (I used a C array where a hashmap would have been better) but I did not by lack of time... I nonetheless think this can be a good idea (I see this repository as a showcase of all you can do with eBPF, so showing you can mix eBPF with almost any language seems a good idea) but I do not represent the whole community so better to get some other feedback before trying.

Thanks for your opinion.
I think it would be better to discuss this issue on `Discussions`
https://github.com/iovisor/bcc/discussions/4355
I hope many community people come and make talk about this issue.

eiffel-fl reviewed (commented):
This tool is still cool and I think merging it would be a good forward step for `libbpf-tools`.
I reviewed the eBPF code itself and it looks OK.
I will nonetheless need to review the userspace code (particularly the one classifying leaks) thoroughly later.

eiffel-fl reviewed (commented):
I did another pass.

Let me summarize to see if I get a good understanding of your code.
Basically, each interval, you interrupt all the threads from the monitored process by using `ptrace()`.
Then in `save_roots()`, you parse the process memory maps to save their begin and end points in different hash maps. I nonetheless do not understand the difference between the `certain` and `uncertain` memory region.
In `read_table()`, you read the eBPF map to know about memory which was allocated but not freed at the moment.
Then in `classify_all_chunks`, you rely a lot on `scan_range_for_pointers()`, which check if it exists pointers to allocated memory block within the given memory area.
You basically calls this function for `uncertain`, `certain` and `registers`. So all the pointers inside these regions will be added to the `frontier` vector.
Then in `flood_fill_tag()`, for each pointer in `frontier`, you scan the memory to which the pointer belongs.
`mark_indirectly_leaked_cb()` is called on each `chunk` and mark the non reachable of them as indirectly leaked.
In `collect_leaks_cb`, you update the reports for directly and indirectly leaked chunks.
You can now detach the process and report the leak to `stdout`.

I have some other comments but I think I have now a better understanding of the C code.

Bojun-Seo commented on libbpf-tools/lsan.bpf.c:
```diff
@@ -0,0 +1,217 @@
+	u64 *size64 = bpf_map_lookup_elem(&sizes, &pid);
+	struct lsan_info_t info = {0, };
+
+	if (size64 == 0)
```
Thanks, I'll apply your suggestion.

Bojun-Seo commented on libbpf-tools/lsan.bpf.c:
```diff
@@ -0,0 +1,217 @@
+int gen_free_enter(struct pt_regs *ctx, void *address)
+{
+	u64 addr = (u64)address;
+	struct lsan_info_t *info = bpf_map_lookup_elem(&allocs, &addr);
+
+	if (info != NULL) {
+		bpf_map_delete_elem(&allocs, &addr);
+	}
```
Your suggestion is reasonable, I'll check and apply it if it works.

Bojun-Seo commented on libbpf-tools/lsan.c:
```diff
@@ -0,0 +1,1331 @@
+	} else
+		p_info("Execute child process: %s", cmd);
+
+	argv = calloc(sizeof(char *), strlen(cmd));
+	if (argv == NULL) {
+		p_err("failed to alloc memory");
+		goto cleanup;
+	}
```
Well, it saves `filepath` seperately. So `argv` should be the number of spaces in `cmd`, I guess.
You are right, it is not appropriate to use `strlen(cmd)` for the size of `argv`.(It wastes memory)
But because of the overhead calculating the number of spaces, I'm thinking about to use `strlen(cmd)/2` instead.
Since the size of `argv` cannot exceed `strlen(cmd)/2` logically and the `argv` will be freed soon.

Bojun-Seo commented on libbpf-tools/lsan.c:
```diff
@@ -0,0 +1,1331 @@
+	}
+
+	/* calloc */
+	func_off = get_elf_func_offset(libc_path, "calloc");
+	if (func_off < 0) {
+		p_err("Failed to get func offset");
+		return func_off;
+	}
+	obj->links.calloc_entry =
+			bpf_program__attach_uprobe(obj->progs.calloc_entry,
+			false, env.pid ? : -1, libc_path, func_off);
+	err = libbpf_get_error(obj->links.calloc_entry);
+	if (err) {
+		p_err("Failed to attach calloc_entry");
+		return err;
+	}
+	obj->links.calloc_return =
+			bpf_program__attach_uprobe(obj->progs.calloc_return,
+			true, env.pid ? : -1, libc_path, func_off);
+	err = libbpf_get_error(obj->links.calloc_return);
+	if (err) {
+		p_err("Failed to attach calloc_return");
+		return err;
+	}
```
I agree with you. In fact, I am trying to use macro function to factorize(and simplify) `attach_uprobes` function. It will be updated soon.

Bojun-Seo commented on libbpf-tools/lsan.c:
```diff
@@ -0,0 +1,1331 @@
+	uptr aa = *(uptr*)a;
+	uptr bb = *(uptr*)b;
+
+	if (aa < bb)
+		return -1;
+
+	if (aa > bb)
+		return 1;
+
+	return 0;
```
I should change my code as you suggested. Thanks!

Bojun-Seo replied:
I realize `return aa - bb` could cause overflow.
Since `uptr` is `unsigned long long` type.
So I think I should retain my code.

Bojun-Seo commented on libbpf-tools/lsan.c:
```diff
@@ -0,0 +1,1331 @@
+		p_warn("Failed to open: %s", env.suppr);
+	} else {
+		while (fgets(line, sizeof(line), fp_suppression) != NULL) {
+			/* suppression line format "kind:string" */
+			/* suppression line example1 "leak:/usr/lib/libglib.so" */
+			/* suppression line example2 "leak:_dl_init" */
```
lsan is a feature came from llvm-project sanitizers.
So I follows the format of llvm-project sanitizers.
Which makes same suppression file can be used on original lsan as well as bcc lsan.

Bojun-Seo commented on libbpf-tools/lsan.c:
```diff
@@ -0,0 +1,1331 @@
+			ptr = strtok(v[MAPS_ADDRESS], "-");
+			begin = strtoull(ptr, NULL, hex);
+			ptr = strtok(NULL, "-");
+			end = strtoull(ptr, NULL, hex);
+			item = (struct root_region_t*)malloc(sizeof(struct root_region_t));
+			ON_MEM_FAILURE(item);
+			item->end = end;
+			item->id = begin;
+			if ('\n' == v[MAPS_PATH][0]) {
+				/* v[MAPS_PATH] value is "\n" if path name does not exist */
+				HASH_FIND(hh, uncertain, &begin, sizeof(uptr), val);
+				if (NULL == val) {
+					item->is_heap = false;
+					HASH_ADD(hh, uncertain, id, sizeof(uptr), item);
+				} else {
+					val->end = end;
+				}
+			} else if (0 == strncmp(heap_str, v[MAPS_PATH], strlen(heap_str))) {
+				HASH_FIND(hh, uncertain, &begin, sizeof(uptr), val);
+				if (NULL == val) {
+					item->is_heap = true;
+					HASH_ADD(hh, uncertain, id, sizeof(uptr), item);
+				} else {
+					val->end = end;
+				}
+			} else if ('[' != v[MAPS_PATH][0] ||
+					0 == strncmp(stack_str, v[MAPS_PATH], strlen(stack_str))) {
+				/* Not including '[' starting path except "[stack]" */
+				HASH_FIND(hh, certain, &begin, sizeof(uptr), val);
+				if (NULL == val) {
+					item->is_heap = false;
+					HASH_ADD(hh, certain, id, sizeof(uptr), item);
+				} else {
+					val->end = end;
+				}
+			}
+		}
+		for (i = 0; i < MAPS_COLUMN_MAX; ++i) {
+			free(v[i]);
+		}
```
Your suggestion seems nice. I'll apply it. Thanks!

Bojun-Seo commented on libbpf-tools/lsan.c:
```diff
@@ -0,0 +1,1331 @@
+
+	set_log_level(INFO);
+#ifdef __arm__
+	p_err("lsan is not supported on arm32 systems");
+	return 0;
```
Are you meaning that it would be better to error at compile time?
Sound reasonable, but I think I should check it with my coworkers.
Since compile error may affect build system in my company.

Bojun-Seo commented:
I would like to explain `certain` and `uncertain` in advance.
`certain` and `uncertain` are maps to store the `start` and `end` point of root regions
The regions stored in `certain` map are certainly root region.(100%, There's no doubt)
The regions stored in `uncertain` map can be a root region but it could be a heap region as well.
So set `is_heap` flag on reading heap addresses. It will be heap region if some heap item address exists between `start` and `end`

Bojun-Seo commented on libbpf-tools/lsan.c:
```diff
@@ -0,0 +1,1331 @@
+	struct root_region_t *next = NULL;
+
+	HASH_ITER(hh, roots, curr, next) {
+		if (!(curr->is_heap)) {
```
Here(function `process_roots`), it targets only memory pointers which are not in the heap.
So if you have a pointer inside the heap, it will not be used in this function.
That pointer inside the heap could be scanned if it is reachable directly from the roots,
and the pointer inside the heap cannot be scanned if it is not reachable directly from the roots.
The pointer inside the heap will be scanned (on `flood_fill_tag` function), if there is a path from the root.

Leak sanitizer algorithm is doing `one-depth DFS` for roots and `BFS` for the `DFS` results.
It uses any value as `node` and pointer as `edge`. (Pointer is a `node` as well as `edge`)
`process_roots` is a function do `one-depth DFS` and `flood_fill_tag` is a function do `DFS`

Bojun-Seo commented on libbpf-tools/lsan.c:
```diff
@@ -0,0 +1,1331 @@
+	if (val == NULL)
+		return;
+
+	if (val->tag == DIRECTLY_LEAKED) {
```
`DIRECTLY_LEAKED` is default value, which is set on lsan.bpf.c:66

Bojun-Seo commented on libbpf-tools/lsan.c:
```diff
@@ -0,0 +1,1331 @@
+	for_each_chunk(collect_ignore_cb);
+	process_roots(certain);
+	process_roots(uncertain);
+	flood_fill_tag(REACHABLE);
```
You are totally right!
In fact, this code pattern comes from original algorithm at `llvm-project lsan`.
I also have been curious about calling `flood_fill_tag` function with `REACHABLE` twice.
I will check the reason of this code.

Thanks again for your review and sorry for late answer.

eiffel-fl reviewed (commented):
Hi!


Thank you for the latest modifications!
I only found nits.
I will test it again but from code inspection it looks OK.


Best regards.

eiffel-fl reviewed (approved):
I tested it again and it works fine:

```bash
$ sudo ./lsan -c /tmp/test -s /tmp/suppr.txt                   (remotes/bojun/lsan) *%
[sudo] Mot de passe de francis : 
Info: Execute child process: /tmp/test
Info: execute command: /tmp/test(pid 94525)
99
99

[2023-11-08 16:15:57] Print leaks:
4 bytes direct leak found in 1 allocations from stack id(41605)
        #1 0x00557fb2fee1bf foo+0x16 (/tmp/test+0x11bf)
        #2 0x00557fb2fee21b main+0x1e (/tmp/test+0x121b)
        #3 0x007f4794629d90 __libc_init_first+0x90 (/usr/lib/x86_64-linux-gnu/libc.so.6+0x29d90)

99

[2023-11-08 16:16:07] Print leaks:
8 bytes direct leak found in 2 allocations from stack id(41605)
        #1 0x00557fb2fee1bf foo+0x16 (/tmp/test+0x11bf)
        #2 0x00557fb2fee21b main+0x1e (/tmp/test+0x121b)
        #3 0x007f4794629d90 __libc_init_first+0x90 (/usr/lib/x86_64-linux-gnu/libc.so.6+0x29d90)

99

[2023-11-08 16:16:17] Print leaks:
12 bytes direct leak found in 3 allocations from stack id(41605)
        #1 0x00557fb2fee1bf foo+0x16 (/tmp/test+0x11bf)
        #2 0x00557fb2fee21b main+0x1e (/tmp/test+0x121b)
        #3 0x007f4794629d90 __libc_init_first+0x90 (/usr/lib/x86_64-linux-gnu/libc.so.6+0x29d90)
```

Apart from the two nits I found earlier, it is, for me, ready to go!
Thank you for this cool tool!

Bojun-Seo commented on libbpf-tools/lsan.c:
```diff
@@ -0,0 +1,1166 @@
+	_attach_uprobe(obj, "valloc", valloc, env.pid);
+	_attach_uprobe(obj, "memalign", memalign, env.pid);
+	_attach_uprobe(obj, "pvalloc", pvalloc, env.pid);
+	_attach_uprobe(obj, "reallocarray", pvalloc, env.pid);
```
Thank you for the catch!
I should change it.

Bojun-Seo commented on libbpf-tools/lsan.c:
```diff
@@ -0,0 +1,1166 @@
+#define _attach_uprobe(obj, func_name_str, func_name, pid)	\
+({	\
+	off_t func_off = get_elf_func_offset_impl(func_name_str);	\
+	__attach_uprobe(obj, func_name##_entry, false, func_off, pid);	\
+	__attach_uprobe(obj, func_name##_return, true, func_off, pid);	\
```
Sorry, I don't quite understand your question. Please let me explain about the probe points. Probe means uprobe in the following context by the way.
`lsan` probes entry point as well as return point of the memory allocation functions(like `malloc`) and probes only entry point of the memory deallocation functions(like `free`), which resembles the tool `memleak`.
The information list of `lsan` wants: `size` and `address` at allocation, `address` at deallocation.
The deallocation `address` can be known at the entry point of the deallocation function. So `lsan` probes the entry point of the deallocation function only.
The allocation `address` can be known at the return point of the allocation function and the `size` can be known at the entry point of the allocation function. So `lsan` probes the entry point and return point of the memory allocation functions.

Bojun-Seo commented on libbpf-tools/lsan.c:
```diff
@@ -0,0 +1,1166 @@
+	func_off;	\
+})
+
+#define __attach_uprobe(obj, func_name, uprobe, func_off, pid)	\
+({	\
+	char *libc_path = GET_LIBC_PATH(); \
+	obj->links.func_name =	\
+		bpf_program__attach_uprobe(obj->progs.func_name, uprobe,	\
+			pid ? : -1, libc_path, func_off);	\
```
I will change the code as you suggested.
Thanks for your meticulous review!

Bojun-Seo commented on libbpf-tools/lsan.c:
```diff
@@ -0,0 +1,1168 @@
+		return rst;
+	}
+	for (i = size - 1; i >= 0; --i) {
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
```
Sounds great, I'll change this code to detect endianness at rutime. Thanks for good idea.

Bojun-Seo replied:
I found that endianess check is not necessary. So I remove it.

Bojun-Seo commented:
Difference between `memleak` and `lsan`
`memleak` shows the status of memory allocation. In other words, it shows memory allocated but not released.
`lsan` shows the unreachable memory among allocated memory.

Followings are the test code
```c
#include <stdlib.h>
#include <unistd.h>

int *arr[10000];

int *foo(size_t size) {
        int *tmp = malloc(size);
        *tmp = 99;
        return tmp;
}

int *bar(size_t nmemb, size_t size) {
        int *tmp = calloc(nmemb, size);
        *tmp = 22;
        return tmp;
}

int *baz(size_t size) {
        int *tmp = valloc(size);
        *tmp = 11;
        return tmp;
}

int main(int argc, char* argv[]) {
        int *a;
        int i = 0;
        while (1) {
                a = foo(4);
                arr[i++] = a;
                a = bar(4, 4);
                free(a);
                a = baz(44);
                sleep(5);
        }
        return 0;
}
```
memory allocated by `foo` is saved on `arr` array. So it is reachable and not deallocated.
memory allocated by `bar` is deallocated.
memory allocated by `baz` will become unreachable(as well as not deallocated).

So, `memleak` reports deallocated allocations. `foo` and `baz`
```
$ sudo ./memleak -c a.out
[sudo] password for bojun:
using default object: libc.so.6
using page size: 4096
tracing kernel: false
child created with pid: 257140
Tracing outstanding memory allocs...  Hit Ctrl-C to end
received go event. executing child command
[14:18:6] Top 2 stacks with outstanding allocations:
88 bytes in 2 allocations from stack
        0 [<000055f3810e1250>] baz+0x1c
        1 [<000055f3810e12d7>] main+0x73
        2 [<00007fbd77829d90>] __libc_init_first+0x90
8 bytes in 2 allocations from stack
        0 [<000055f3810e11e5>] foo+0x1c
        1 [<000055f3810e1288>] main+0x24
        2 [<00007fbd77829d90>] __libc_init_first+0x90
[14:18:11] Top 2 stacks with outstanding allocations:
132 bytes in 3 allocations from stack
        0 [<000055f3810e1250>] baz+0x1c
        1 [<000055f3810e12d7>] main+0x73
        2 [<00007fbd77829d90>] __libc_init_first+0x90
12 bytes in 3 allocations from stack
        0 [<000055f3810e11e5>] foo+0x1c
        1 [<000055f3810e1288>] main+0x24
        2 [<00007fbd77829d90>] __libc_init_first+0x90

^C[14:18:12] Top 2 stacks with outstanding allocations:
132 bytes in 3 allocations from stack
        0 [<000055f3810e1250>] <null sym>
        1 [<000055f3810e12d7>] <null sym>
        2 [<00007fbd77829d90>] <null sym>
12 bytes in 3 allocations from stack
        0 [<000055f3810e11e5>] <null sym>
        1 [<000055f3810e1288>] <null sym>
        2 [<00007fbd77829d90>] <null sym>
reaped child process
done
```

`lsan` reports only unreachable allocations. `baz`.
```
$ sudo ./lsan -c a.out
2024-Mar-14 11:47:18 INFO execute command: a.out(pid 257157)

[2024-03-14 14:26:51] Print leaks:
44 bytes direct leak found in 1 allocations from stack id(65376)
        #1 0x00559d087a2250 baz+0x1c
        #2 0x00559d087a22d7 main+0x73
        #3 0x007f0b3a629d90 [unknown]


[2024-03-14 14:27:01] Print leaks:
132 bytes direct leak found in 3 allocations from stack id(65376)
        #1 0x00559d087a2250 baz+0x1c
        #2 0x00559d087a22d7 main+0x73
        #3 0x007f0b3a629d90 [unknown]

^C
```

eiffel-fl reviewed (commented):
Hi!

I already approved it, so it should be OK.
Reading the code again, I just have one question with regard to avoid having several loops, but I am doubtful this is possible.

Best regards.

Bojun-Seo commented on libbpf-tools/lsan.c:
```diff
@@ -0,0 +1,1157 @@
+	cvector_reserve(tids, 1024);
+	len = get_tids(env.pid, cvector_begin(tids), 1024);
+	cvector_set_size(tids, len);
+	for_each_tid_ptrace(PTRACE_SEIZE);
+	for_each_tid_ptrace(PTRACE_INTERRUPT);
+	for_each_tid_waitpid();
```
Yes, it could be possible to have the code inside one loop and doing all `ptrace` `attach`, `interrupt` and `waitpid` inside the loop, which will mitigate loop overhead.

I think it would be better to factorize it for readability.
However, despite the intention to factorize it for readability, it currently doesn't seem to be very readable.
I'll revise this function to enhance its readability.

Thanks!

Bojun-Seo replied:
I tried to change the code as you suggested, but because of the `env.stop_the_world` flag, the code became more complicated. This means the current code has better readability in my opinion. So I decided to stick with it. However, your suggestion is very good, and I'll change the code if I find a more readable way.

Thanks!

Bojun-Seo committed:
libbpf-tools: Add submodule c-vector

Add submodule c-vector
https://github.com/eteran/c-vector

This project is header only library project that contains interfaces
and implementations of vector container which can be used in C
The license of this project is MIT

Bojun-Seo committed:
libbpf-tools: Add submodule uthash

Add submodule uthash
https://github.com/troydhanson/uthash

This project is header only library project that contains interfaces
and implementations of hash map container which can be used in C
The license is written on the top of the file uthash.h

Bojun-Seo committed:
libbpf-tools: Add new feature ALSan on libbpf-tools

Add ALSan(Attachable Leak Sanitizer) feature on libbpf-tools
ALSan feature originally comes from the llvm-project lsan
https://github.com/llvm/llvm-project
This tool detect and report unreachable memory periodically

USAGE:

  $ ./alsan -h
  Usage: alsan [OPTION...]
  Detect memory leak resulting from unreachable pointers.

  Either -c or -p is a mandatory option
  EXAMPLES:
      alsan -p 1234             # Detect leaks on process id 1234
      alsan -c a.out            # Detect leaks on a.out
      alsan -c 'a.out arg'      # Detect leaks on a.out with argument

    -c, --command=COMMAND      Execute and detect memory leak on the specified
                               command
    -i, --interval=INTERVAL    Set interval in second to detect leak
    -p, --pid=PID              Detect memory leak on the specified process
    -s, --suppressions=SUPPRESSIONS
                               Suppressions file name
    -T, --top=TOP              Report only specified amount of backtraces
    -v, --verbose              Verbose debug output
    -w, --stop-the-world       Stop the target process during tracing
    -?, --help                 Give this help list
        --usage                Give a short usage message
    -V, --version              Print program version

  Mandatory or optional arguments to long options are also mandatory or optional
  for any corresponding short options.

  Report bugs to https://github.com/iovisor/bcc/tree/master/libbpf-tools.

Report example:

  $ sudo ./alsan -p 28346

  [2024-05-22 14:44:58] Print leaks:
  44 bytes direct leak found in 1 allocations from stack id(57214)
          #1 0x00583bca1b2250 baz+0x1c (/home/bojun/alsan/libbpf-tools/a.out+0x1250)
          #2 0x00583bca1b22d7 main+0x73 (/home/bojun/alsan/libbpf-tools/a.out+0x12d7)
          #3 0x007470c7c2a1ca [unknown] (/usr/lib/x86_64-linux-gnu/libc.so.6+0x2a1ca)
          #4 0x007470c7c2a28b __libc_start_main+0x8b (/usr/lib/x86_64-linux-gnu/libc.so.6+0x2a28b)
          #5 0x00583bca1b2105 _start+0x25 (/home/bojun/alsan/libbpf-tools/a.out+0x1105)

  [2024-05-22 14:45:08] Print leaks:
  132 bytes direct leak found in 3 allocations from stack id(57214)
          #1 0x00583bca1b2250 baz+0x1c (/home/bojun/alsan/libbpf-tools/a.out+0x1250)
          #2 0x00583bca1b22d7 main+0x73 (/home/bojun/alsan/libbpf-tools/a.out+0x12d7)
          #3 0x007470c7c2a1ca [unknown] (/usr/lib/x86_64-linux-gnu/libc.so.6+0x2a1ca)
          #4 0x007470c7c2a28b __libc_start_main+0x8b (/usr/lib/x86_64-linux-gnu/libc.so.6+0x2a28b)
          #5 0x00583bca1b2105 _start+0x25 (/home/bojun/alsan/libbpf-tools/a.out+0x1105)

Source code of test program:

  $ cat leak_test.c
  #include <stdlib.h>
  #include <unistd.h>

  int *arr[10000];

  int *foo(size_t size) {
          int *tmp = malloc(size);
          *tmp = 99;
          return tmp;
  }

  int *bar(size_t nmemb, size_t size) {
          int *tmp = calloc(nmemb, size);
          *tmp = 22;
          return tmp;
  }

  int *baz(size_t size) {
          int *tmp = valloc(size);
          *tmp = 11;
          return tmp;
  }

  int main(int argc, char* argv[]) {
          int *a;
          int i = 0;
          while (1) {
                  a = foo(4);
                  arr[i++] = a;
                  a = bar(4, 4);
                  free(a);
                  a = baz(44);
                  sleep(5);
          }
          return 0;
  }

Bojun-Seo commented:
I split my whole patch into three parts.
1. Adding open source cvector.h file
2. Adding open source uthash.h file
3. Lsan implementation

Bojun-Seo commented:
I changed `alignment` value(loop stride factor) in function `scan_range_for_pointers` from `1` to `WORD_SIZE` for performance.

eiffel-fl reviewed (commented):
Hi!

I have one comment with regard to `cvector.h` and `uthash.h`.

Best regards.

Bojun-Seo commented on libbpf-tools/cvector.h:
```diff
@@ -0,0 +1,294 @@
+/*
```
I changed the commits to use `git submodule`.
Always thanks for your great review!

Bojun-Seo commented:
Rebase and change the tool name from `lsan` to `alsan` because `lsan` is a name that is already in use.
This tool is not `lsan`, so I changed the name to `alsan` to make it clear.

Bojun-Seo commented:
There was a talk about this feature in C++ Now conference.

https://youtu.be/9f5hd-8suVE?si=lOli99myFpExgzac

I hope this video helps people understand this patch.
